fetcher.go

```diff
@@ -35,6 +37,27 @@ type tokenFetcher struct {
 	grantType    string
 	tokenURL     string
 	httpClient   *http.Client
+	cachedToken  string
+}
+
+// Returns the cached token, if that has expired or does not exist it returns a new token
+func (a *tokenFetcher) CachedToken(audience string) (string, error) {
+	if a.cachedToken != "" {
+		var p jwt.Parser
+		// Check expiration of token, this does not need to be verified because
+		// verification occurs on the server.
+		token, _, _ := p.ParseUnverified(a.cachedToken, &jwt.StandardClaims{})
+		if token != nil && token.Claims.Valid() == nil {
+			return a.cachedToken, nil
+		}
+	}
+
+	t, err := a.Token(audience)
+	if err != nil {
+		return "", err
+	}
+	a.cachedToken = t
+	return a.cachedToken, nil
```

Comment:
I'd rename this to Token() and change the existing Token() function to newToken() (private). As a user I just want a token, I don't care if it's cached or not. Also, existing code may not need to updated to use the new library.